src/java/org/apache/cassandra/index/sai/disk/v1/TermsReader.java

```diff
@@ -203,6 +214,96 @@ public PostingsReader getPostingReader(long offset) throws IOException
+        private Expression exp;
+
+        RangeQuery(Expression exp, QueryEventListener.TrieIndexEventListener listener, QueryContext context)
+        {
+            this.listener = listener;
+            postingsInput = IndexFileUtils.instance.openInput(postingsFile);
+            postingsSummaryInput = IndexFileUtils.instance.openInput(postingsFile);
+            this.exp = exp;
+            lookupStartTime = System.nanoTime();
+            this.context = context;
+        }
+
+        public PostingList execute()
+        {
+            final ByteComparable lower = exp.lower != null ? ByteComparable.fixedLength(exp.getLowerBound()) : null;
+            final ByteComparable upper = exp.upper != null ? ByteComparable.fixedLength(exp.getUpperBound()) : null;
+            try (TrieRangeIterator reader = new TrieRangeIterator(termDictionaryFile.instantiateRebufferer(),
+                                                                  termDictionaryRoot,
+                                                                  lower,
+                                                                  upper,
+                                                                  exp.lowerInclusive,
+                                                                  exp.upperInclusive))
+            {
+                var iter = reader.iterator();
+                if (!iter.hasNext())
+                {
+                    closeQuietly();
+                    return PostingList.EMPTY;
+                }
+
+                context.checkpoint();
+                // Because postings are not sorted, we need to eagerly materialize the results and sort them.
+                LongHeap postings = materializeResults(iter);
+                return new LongHeapPostingList(postings);
+            }
+            catch (Throwable e)
+            {
+                //TODO Is there an equivalent of AOE in OS?
+                if (!(e instanceof AbortedOperationException))
+                    logger.error(indexContext.logMessage("Failed to execute term query"), e);
+
+                closeQuietly();
+                throw Throwables.cleaned(e);
+            }
+        }
+
+        /**
+         * Build an in-memory heap of row ids from the posting lists of the matching terms.
+         * @return an ordered {@link LongHeap} of row ids
+         */
+        private LongHeap materializeResults(Iterator<Pair<ByteSource,Long>> iterable) throws IOException
+        {
+            var heap = new LongHeap(1);
+            while (iterable.hasNext())
+            {
+                Pair<ByteSource,Long> next = iterable.next();
+                byte[] nextBytes = ByteSourceInverse.readBytes(next.left);
+                if (exp.isSatisfiedBy(ByteBuffer.wrap(nextBytes)))
+                {
+                    // TODO is this the correct way to get the posting list?
+                    var currentReader = new PostingsReader(postingsInput, next.right, listener.postingListEventListener());
```

Comment:
Good question. The closing the `PostingsReader` only closes the `postingsInput`, and because we're sharing it across multiple, I didn't close it. This seems pretty error prone, so I'll revisit it.

Reply:
I updated the logic to make the contract clearer, please take a look.